File: Algorithm/QuickSort.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        // write your code here
        int[] arr = {3,5,2,6,4,7,9,1,8};

        for(int num: arr){
            System.out.printf("%d, ", num);
        }
        System.out.println();

        quickSort(arr, 0, arr.length-1);

        for(int num: arr){
            System.out.printf("%d, ", num);
        }
    }

    public static void quickSort(int[] arr, int left, int right){
        if(left >= right) return;

        int pivot = partition(arr, left, right);

        quickSort(arr, left, pivot);
        quickSort(arr, pivot+1, right);
    }

    public static int partition(int[] arr, int left, int right){
        int start = left, end = right;
        int pivot = arr[(left + right) / 2];
        while(start < end) {
            while(arr[start] < pivot) {
                start++;
            }
            while(arr[end] > pivot){
                end--;
            }
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
        }

        return start;
    }
}

File: Algorithm/Udemy_Javascript/sort/quickSort.js
function partition(arr, start = 0, end = arr.length - 1) {
  let pivot = arr[start];
  let pivotIdx = start;
  for (let i = start + 1; i <= end; i++) {
    if (pivot > arr[i]) {
      pivotIdx++;
      [arr[pivotIdx], arr[i]] = [arr[i], arr[pivotIdx]];
    }
  }
  [arr[pivotIdx], arr[start]] = [arr[start], arr[pivotIdx]];

  return pivotIdx;
}

function quickSort(arr, left = 0, right = arr.length - 1) {
  if (left < right) {
    let pivot = partition(arr, left, right);

    quickSort(arr, left, pivot - 1);
    quickSort(arr, pivot + 1, right);
  }

  return arr;
}

console.log(quickSort([4, 6, 9, 1, 2, 5, 3]));
